dbms/src/Interpreters/JoinV2/HashJoinKey.h

```diff
@@ -0,0 +1,691 @@
+    size_t required_key_offset = 0;
+};
+
+class alignas(CPU_CACHE_LINE_SIZE) HashJoinKeysFixedOther
+{
+public:
+    using KeyType = StringRef;
+    explicit HashJoinKeysFixedOther(const TiDB::TiDBCollators &) {}
+
+    void reset(const ColumnRawPtrs & key_columns, size_t raw_required_key_size)
+    {
+        RUNTIME_CHECK(!key_columns.empty());
+        rows = key_columns[0]->size();
+        size_t sz = key_columns.size();
+        vec.resize(sz);
+        fixed_size.resize(sz);
+        fixed_size_sum = 0;
+        for (size_t i = 0; i < sz; ++i)
+        {
+            vec[i] = key_columns[i]->getRawData().data;
+            fixed_size[i] = key_columns[i]->sizeOfValueIfFixed();
+            fixed_size_sum += fixed_size[i];
+        }
+        key_buffer.resize(fixed_size_sum);
+
+        RUNTIME_CHECK(raw_required_key_size <= sz);
+        required_key_offset = fixed_size_sum;
+        for (size_t i = 0; i < raw_required_key_size; ++i)
+            required_key_offset -= fixed_size[sz - 1 - i];
+
+        buffer_initialized = false;
+    }
+
+    ALWAYS_INLINE KeyType getJoinKey(size_t row)
+    {
+        size_t sz = vec.size();
+        size_t offset = 0;
+        for (size_t i = 0; i < sz; ++i)
+        {
+            switch (fixed_size[i])
+            {
+            case 1:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 1);
+                break;
+            case 2:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 2);
+                break;
+            case 4:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 4);
+                break;
+            case 8:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 8);
+                break;
+            case 16:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 16);
+                break;
+            case 32:
+                tiflash_compiler_builtin_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, 32);
+                break;
+            default:
+                inline_memcpy(&key_buffer[offset], vec[i] + fixed_size[i] * row, fixed_size[i]);
+            }
+            offset += fixed_size[i];
+        }
+        return StringRef(key_buffer.data(), fixed_size_sum);
+    }
+
+    ALWAYS_INLINE KeyType getJoinKeyWithBuffer(size_t row)
+    {
+        if unlikely (!buffer_initialized)
+        {
+            keys_buffer.resize(rows * fixed_size_sum);
+            size_t offset = 0;
+            size_t sz = vec.size();
+            for (size_t i = 0; i < rows; ++i)
+            {
+                for (size_t j = 0; j < sz; ++j)
+                {
+                    inline_memcpy(&keys_buffer[offset], vec[j] + fixed_size[j] * i, fixed_size[j]);
+                    offset += fixed_size[j];
+                }
```

Comment:
forget to set `buffer_initialized` to true?

Reply:
Nice catch. Fixed.